### PR TITLE
Add perf test script using avalanche

### PR DESCRIPTION
**Pull request by pmalek-sumo**
###### Description

Add a script to run performance test in vagrant with metrics (using [avalanche](https://github.com/open-fresh/avalanche) as metrics generator).

```
Usage:
TEST_DURATION=<duration_in_seconds> TEST_WARMUP_DURATION=<warmup_in_seconds> ./vagrant/scripts/perf-test.sh
```

###### Testing performed

- [ ] ci/build.sh
- [ ] Redeploy fluentd and fluentd-events pods
- [ ] Confirm events, logs, and metrics are coming in

### CODE REVIEW

File: vagrant/k8s/avalanche.yaml [hunk 1]
```diff
@@ -29,17 +28,14 @@ spec:
     spec:
       containers:
       - name: avalanche
-        image: quay.io/freshtracks.io/avalanche:master-2019-10-09-3efdac3
+        image: quay.io/freshtracks.io/avalanche:latest
```
Comment (sumo-drosiek) [inline vagrant/k8s/avalanche.yaml:31]: I think we should use fixed version

Comment (pmalek-sumo) [reply vagrant/k8s/avalanche.yaml]: Done

File: vagrant/scripts/perf-test.sh [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+# Letting prometheus ingest metrics for 20s...
+# Cleaning up avalanche...
+#
+# Prometheus container memory RSS:
+#         start:           657Mi
+#         finish:         1200Mi
+#         diff:            543Mi
+# receiver-mock ingested 3575 metrics during the test
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:26]:
```suggestion
DIR="$( realpath "$( dirname "${BASH_SOURCE[0]}" )" )"
```

File: vagrant/scripts/perf-test.sh [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+AVALANCHE_YAML_PATH="${DIR}/../k8s/avalanche.yaml"
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:27]:
```suggestion
readonly AVALANCHE_YAML_PATH="${DIR}/../k8s/avalanche.yaml"
```

File: vagrant/scripts/perf-test.sh [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+readonly RECEIVER_MOCK_POD="$(kubectl get pod -n receiver-mock --no-headers -lapp=receiver-mock --output custom-columns=NAME:.metadata.name)"
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:28]: what if there will be more than one receiver-mock pod

File: vagrant/scripts/perf-test.sh [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+function is_avalanche_running() {
+    out="$(kubectl get pods --no-headers -n avalanche --output custom-columns=NAME:.metadata.name)"
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:31]: make `out` local

File: vagrant/scripts/perf-test.sh [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        -- curl -s -g -XPOST collection-prometheus-oper-prometheus.sumologic.svc.cluster.local:9090/api/v1/query \
+        --data-urlencode "query=container_memory_rss{job='kubelet',container='${container_name}'}" \
+        | jq -r '.data.result[0].value[1]'
+}
+
+function ingested_metrics_count() {
+    local count
+    count="$(kubectl exec -n receiver-mock "${RECEIVER_MOCK_POD}" \
+        -- curl -s -XGET localhost:3000/metrics | grep "^receiver_mock_metrics_count" | awk '{print $2}')"
+    echo "${count}"
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:52]: can we just do `kubectl exec`? It will print result on the stdout anyway

File: vagrant/scripts/perf-test.sh [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    kubectl exec -n receiver-mock "${RECEIVER_MOCK_POD}" \
+        -- curl -s -XPOST localhost:3000/metrics-reset >/dev/null
+}
+
+function run_test() {
+    local duration_seconds="${1}"
+    echo "Starting perf tests"
+
+    if [[ $(is_avalanche_running) -eq 0 ]]; then
+        echo "Avalanche is running, stop it, wait for the collection to cool down and then rerun the test"
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:65]: Don't we want to test collection in hot state? I would like to avoid testing during bootstraping

Comment (sumo-drosiek) [reply vagrant/scripts/perf-test.sh:65]: I mean, it's fine for prometheus but doesn't give us top values for the DPM

Comment (pmalek-sumo) [reply vagrant/scripts/perf-test.sh:65]: What do you suggest then? To not check if avalanche is running when the test is started?

File: vagrant/scripts/perf-test.sh (hunk 2 above)
Comment (pmalek-sumo) [reply vagrant/scripts/perf-test.sh:26]: 🤔

File: vagrant/scripts/perf-test.sh [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    echo "Cleaning up avalanche..."
+    kubectl delete -f "${AVALANCHE_YAML_PATH}" >/dev/null ||
+        (echo "Failed deleting avalanche yaml (${AVALANCHE_YAML_PATH})" && exit 1)
+
+    printf "\n"
+    printf "Prometheus container memory RSS:\n"
+    printf "\tstart:\t%12dMi\n" "$(( PROMETHEUS_MEMORY_RSS_START / 1024 / 1024 ))"
+    printf "\tfinish:\t%12dMi\n" "$(( PROMETHEUS_MEMORY_RSS_FINISH / 1024 / 1024 ))"
+    printf "\tdiff:\t%12dMi\n" "$(( ("${PROMETHEUS_MEMORY_RSS_FINISH} - ${PROMETHEUS_MEMORY_RSS_START}") / 1024 / 1024 ))"
+}
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:95]: Make the vars local

File: vagrant/scripts/perf-test.sh [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        echo "Avalanche is running, stop it, wait for the collection to cool down and then rerun the test"
+        exit 1
+    fi
+
+    PROMETHEUS_MEMORY_RSS_START=$(container_memory_rss prometheus)
+
+    echo "Setting up avalanche..."
+    kubectl apply --wait -f "${AVALANCHE_YAML_PATH}" >/dev/null || \
+        (echo "Failed applying avalanche yaml (${AVALANCHE_YAML_PATH})" && exit 1)
+    kubectl wait --timeout=120s -n avalanche --for=condition=Available deployment/avalanche >/dev/null
```
Comment (sumo-drosiek) [inline vagrant/scripts/perf-test.sh:73]: Shouldn't we escape space between `Available` and `deployment`?

File: vagrant/scripts/perf-test.sh (hunk 7 above)
Comment (sumo-drosiek) [reply vagrant/scripts/perf-test.sh:64]: At first comment i didn't though about prometheus, so I think we should do as follow

1. check is cluster is cold
2. measure prometheus metrics
3. wait until cluster is hot
4. clean up receiver-mock metrics
5. measure receiver-mock ingest
6. wait some time
7. measure receiver-mock ingest
8. measure prometheus metrics

Comment (sumo-drosiek) [reply vagrant/scripts/perf-test.sh:64]: Does it make sense for you?

File: vagrant/scripts/perf-test.sh (hunk 9 above)
Comment (pmalek-sumo) [reply vagrant/scripts/perf-test.sh:73]: Hmm no, why? even `kubectl wait --help` has a similar example:

```
Examples:
  # Wait for the pod "busybox1" to contain the status condition of type "Ready".
  kubectl wait --for=condition=Ready pod/busybox1
```

File: vagrant/scripts/perf-test.sh (hunk 4 above)
Comment (pmalek-sumo) [reply vagrant/scripts/perf-test.sh]: added `| head -1` :)

File: vagrant/scripts/perf-test.sh (hunk 7 above)
Comment (pmalek-sumo) [reply vagrant/scripts/perf-test.sh:64]: What's the condition for "cold" and "hot" cluster?

We can add second parameter to this script being the warmup period

File: vagrant/scripts/perf-test.sh (hunk 4 above)
Comment (sumo-drosiek) [reply vagrant/scripts/perf-test.sh]: Thats fine for executing but not for counting the total metrics 😅

File: vagrant/scripts/perf-test.sh (hunk 9 above)
Comment (sumo-drosiek) [reply vagrant/scripts/perf-test.sh:78]: :facepalm: thats fine

File: vagrant/scripts/perf-test.sh (hunk 7 above)
Comment (sumo-drosiek) [reply vagrant/scripts/perf-test.sh:64]: I would say 1 min for warmup is enough

Comment (pmalek-sumo) [reply vagrant/scripts/perf-test.sh:75]: Please check if current version makes sense

File: deploy/helm/sumologic/templates/setup/setup-custom-configmap.yaml [hunk 10]
```diff
@@ -9,6 +8,7 @@ metadata:
     app: {{ template "sumologic.labels.app.setup.configmap-custom" . }}
     {{- include "sumologic.labels.common" . | nindent 4 }}
 data:
+{{- if .Values.sumologic.setup.additionalFiles }}
```
Comment (perk-sumo) [inline deploy/helm/sumologic/templates/setup/setup-custom-configmap.yaml:11]: Why this change? It should be a separate PR if we need it IMO.

Comment (pmalek-sumo) [reply deploy/helm/sumologic/templates/setup/setup-custom-configmap.yaml]: Just merged master. Just be good now

**Review by perk-sumo — APPROVED** [review]
👍

File: vagrant/Makefile [hunk 11]
```diff
@@ -73,12 +74,34 @@ apply-service-monitors:
...
+# This applies prometheus remote write rule to forward all avalanche metrics to receiver-mock
+# Since this removes the generated values.local.yaml after it's applied subsequent 'make helm-upgrade'
+# will revert this change.
+apply-avalanche-remote-write: _apply-avalanche-remote-write-to-local-values helm-upgrade clean-local-values-yaml
+
+clean-local-values-yaml:
+	rm ${local_values_file}
+
+_apply-avalanche-remote-write-to-local-values: generate-local-config
+	yq w -s ${vagrant_scripts_dir}yq/update_remote_write_for_avalanche.yaml ../deploy/helm/sumologic/values.yaml | \
```
Comment (sumo-drosiek) [inline vagrant/Makefile:86]: Where is this (`update_remote_write_for_avalanche.yaml`) file? 🤔

Comment (pmalek-sumo) [reply vagrant/Makefile:86]: Ups, added in https://github.com/SumoLogic/sumologic-kubernetes-collection/pull/1025/commits/acac7d58f006611b08c621239b93c6ae138cf4f2

File: vagrant/scripts/yq/update_remote_write_for_avalanche.yaml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+- command: update
+  path: prometheus-operator.prometheus.prometheusSpec.remoteWrite[+]
+  value:
+    url: http://$(CHART).$(NAMESPACE).svc.cluster.local:9888/prometheus.metrics.avalanche
+    writeRelabelConfigs:
+    - action: keep
+      regex: avalanche.*
+      sourceLabels: [__name__]
```
Comment (sumo-drosiek) [inline vagrant/scripts/yq/update_remote_write_for_avalanche.yaml:8]:
```suggestion
      sourceLabels: [__name__]

```

**Review by sumo-drosiek — APPROVED** [review]
🏎️
